### ISSUE TITLE
encoding/json: UnmarshalTypeError.Error message seems insensible

### ISSUE THREAD

**Issue by dsnet** [golang/go#43126]
[Consider this snippet](https://play.golang.org/p/lKh1SHb-9VT):
```go
type GoStruct1 struct {
	AlphaField GoStruct2 `json:"alpha_field"`
}
type GoStruct2 struct {
	BravoField []GoStruct3 `json:"bravo_field"`
}
type GoStruct3 struct {
	StringField string `json:"string_field"`
}

func main() {
	fmt.Println(json.Unmarshal([]byte(`{"alpha_field":{"bravo_field":[{"string_field": 3.14159}]}}`), new(GoStruct1)))
}
```

On Go1.15, this prints:
```
json: cannot unmarshal number into Go struct field GoStruct3.alpha_field.bravo_field.string_field of type string
```
Notice that it combines two different namespaces where it is:
```
{{GoTypeName}}.{{JSONPath}}
```
and that the message says that it is a "Go struct field" when it isn't.
Also, it uses `GoStruct3`, when it makes little sense.
At minimum, it should at least used `GoStruct1`.

\cc @mvdan

**Comment by dsnet** [comment]
Related considerations:
* Should the JSON path include array indexes?
* Should the JSON path disambiguate the path if dots appear in the object name? [Example](https://play.golang.org/p/ZkMvETBcFHa).

Strawman proposal: the JSON path should be a valid expression in JavaScript to address the specific value from the root value.
* For objects, we use `.foo` if the object name is a valid JavaScript identifier, otherwise we use `["fizz.buzz"]` if it isn't.
* For arrays, we use `[5]`.

**Comment by mvdan** [comment]
I agree with the strawman proposal. In that proposal, what would you use in place of GoStruct3?

Also, is there a "json path" spec we can simply adhere to? or do we simply try to get as close to a JS expression as we reasonably can?

cc @cuonglm who worked on the original feature. I did think about proposing to make it more consistent with slices and such, but at the end of the day I thought the addition was already a big step forward.

**Comment by dsnet** [comment]
> Also, is there a "json path" spec we can simply adhere to? 

One option is [RFC 6901](https://tools.ietf.org/html/rfc6901), which uses `/` instead of `.` as the delimiter. This would be the visual difference between `.fooName["quoted/name"].barName[5]` and `/fooName/quoted~1Name/barName/5`. I kind of like the simplicity and visual aesthetics of this. It only requires escaping of the `~` and `/` as `~0` and `~1` respectively.

> do we simply try to get as close to a JS expression as we reasonably can?

The grammar for a JS identifier is pretty complicated. Fortunately, the Go identifier grammar is a strict subset of the JS identifier grammar, so we could just use that as a shortcut.

**Comment by dsnet** [comment]
Another oddity. If the target type is a Go map, then the JSON object names for those are dropped: https://play.golang.org/p/q9Yg7Ng01G9.

**Comment by fl0cke** [comment]
The `GoStruct3` at the beginning is extremely confusing. I would either

- replace it with the type name of the outermost struct, e.g. `GoStruct1.alpha_field.bravo_field.string_field`
- replace it with a path expression that is similar to the one used by MariaDB and MySQL, e.g. `$.alpha_field.bravo_field.string_field` (the $ symbol represents the start of the path context). See https://mariadb.com/kb/en/jsonpath-expressions/. 
- omit the first first completely, e.g. `alpha_field.bravo_field.string_field`

**Comment by dsnet** [comment]
The v1 error messages are more or less frozen for backward compatibility reasons. With the release of Go 1.27 and encoding/json/v2, we consistently always use JSON Pointer (RFC 6901) to report the position of a problem.
